scripts/enforce_imports.py

```diff
@@ -0,0 +1,178 @@
+        self._check_split_import(node)
+        self._check_order(node)
+
+    def visit_ImportFrom(self, node: ast.ImportFrom) -> None:
+        self._check_split_import(node)
+        self._check_no_parentheses(node)
+        self._check_order(node)
+
+    def _check_split_import(self, node: ast.AST) -> None:
+        if len(node.names) > 1:
+            self._error(node, Errors.import_not_split)
+
+    def _check_no_parentheses(self, node: ast.AST) -> None:
+        if any(token.string == '(' for token in self._tokenize_line(node.lineno)):
+            self._error(node, Errors.import_with_parentheses)
+
+    def _check_order(self, node: ast.AST) -> None:
+        order_info = self._order_info(node)
+        if order_info is None:
+            return
+        # The order in which NodeVisitor traverses the syntax tree is unspecified
+        # so we can't be sure which nodes have already been visited.
+        # To prevent a single out-of-order import from causing errors on every
+        # other line, we only compare each import statement with one other statement.
+        # To prevent the same line from being reported twice, the error is always
+        # reported for the node we're currently visiting, regardless of whether it
+        # comes first or second.
+        pred, succ = self._visited_predecessor(node.lineno), self._visited_successor(node.lineno)
+        if pred is not None and not self._is_correct_order(pred[0].lineno, node.lineno, pred[1], order_info):
+            self._error(node, Errors.import_not_ordered)
+        elif succ is not None and not self._is_correct_order(node.lineno, succ[0].lineno, order_info, succ[1]):
+            self._error(node, Errors.import_not_ordered)
+        self.visited.append((node, order_info))
+
+    def _error(self, node: ast.AST, err: Errors) -> None:
+        self.errors.append((node.lineno, node.col_offset, err.value))
+
+    def _is_correct_order(self, lineno1: int, lineno2: int, order_info1: ModuleOrderInfo, order_info2: ModuleOrderInfo):
+        return (lineno1 < lineno2) == (order_info1 < order_info2)
+
+    def _order_info(self, node: ast.AST) -> Optional[ModuleOrderInfo]:
+        try:
+            name = one(node.names).name
+        except ValueError:
+            # Don't bother checking order on imports that aren't split
+            return None
+        try:
+            module_name = node.module
+            target_name = name
+        except AttributeError:
+            module_name = name
+            target_name = ''
+        return ModuleType.from_module_name(module_name), module_name, target_name
+
+    def _visited_successor(self, lineno: int) -> Optional[ast.AST]:
+        return min(filter(lambda t: t[0].lineno > lineno, self.visited),
+                   key=lambda t: t[0].lineno,
+                   default=None)
+
+    def _visited_predecessor(self, lineno: int) -> Optional[ast.AST]:
+        return max(filter(lambda t: t[0].lineno < lineno, self.visited),
```

Comment:
I'm not making any assumptions. `ast.NodeVisitor` does not define the order in which nodes are visited, so when we process an import statement we cannot know which (if any) other imports have already been visited. In practice it seems to go from to top to bottom but I can't find any documentation that requires this. I tried to design the code so that it would be eventually correct even if the nodes are visited in completely random order.